File: packages/python/elizaos/features/advanced_capabilities/experience/actions/record_experience.py
from __future__ import annotations

import logging
import time
from typing import TYPE_CHECKING
from uuid import uuid4

from elizaos.types import Action, ActionResult

if TYPE_CHECKING:
    from elizaos.types import (
        HandlerCallback,
        HandlerOptions,
        IAgentRuntime,
        Memory,
        State,
    )

logger = logging.getLogger("elizaos.experience")


async def _validate_record_experience(
    runtime: IAgentRuntime,
    message: Memory,
    _state: State | None = None,
) -> bool:
    """Validate that the message is requesting to record an experience."""
    text = ""
    if hasattr(message, "content") and message.content:
        text_val = getattr(message.content, "text", None)
        if isinstance(text_val, str):
            text = text_val.lower()

    if not text:
        return False

    # Check for relevant keywords
    return "remember" in text or "record" in text or "experience" in text


async def _handle_record_experience(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
    options: HandlerOptions | None = None,
    callback: HandlerCallback | None = None,
    responses: list[Memory] | None = None,
) -> ActionResult | None:
    """Record an experience from the current message context."""
    _ = options, callback, responses  # Unused parameters

    logger.info("Recording experience for message: %s", getattr(message, "id", "unknown"))

    # Create experience memory with context
    experience_id = str(uuid4())
    message_text = ""
    message_source = ""
    if hasattr(message, "content") and message.content:
        text_val = getattr(message.content, "text", None)
        if isinstance(text_val, str):
            message_text = text_val
        source_val = getattr(message.content, "source", None)
        if isinstance(source_val, str):
            message_source = source_val

    state_text = ""
    if state and hasattr(state, "text"):
        state_text = str(getattr(state, "text", ""))

    experience_memory = {
        "id": experience_id,
        "entity_id": str(getattr(message, "entity_id", "")),
        "agent_id": str(runtime.agent_id),
        "room_id": str(getattr(message, "room_id", "")),
        "content": {
            "text": message_text,
            "source": message_source,
            "type": "experience",
            "context": state_text,
        },
        "created_at": int(time.time() * 1000),
    }

    # Store in experiences table
    await runtime.create_memory(experience_memory, "experiences", True)
    logger.info("Experience recorded successfully")

    return ActionResult(
        success=True,
        text="Experience recorded.",
        data={"experienceMemoryId": experience_id},
    )


record_experience_action = Action(
    name="RECORD_EXPERIENCE",
    description="Records an experience or learning from the current conversation context for future reference",
    handler=_handle_record_experience,
    validate=_validate_record_experience,
    similes=["remember experience", "save learning", "record learning"],
    examples=[],
)

File: packages/typescript/src/features/advanced-capabilities/experience/actions/record-experience.ts
import { createUniqueUuid } from "../../../../entities.ts";
import { logger } from "../../../../logger.ts";
import type {
	Action,
	ActionExample,
	ActionResult,
	HandlerCallback,
	HandlerOptions,
} from "../../../../types/components.ts";
import type { Memory } from "../../../../types/memory.ts";
import type { IAgentRuntime } from "../../../../types/runtime.ts";
import type { State } from "../../../../types/state.ts";
import { requireActionSpec } from "../generated/specs/spec-helpers";

const spec = requireActionSpec("RECORD_EXPERIENCE");

export const recordExperienceAction: Action = {
	name: spec.name,
	similes: spec.similes ? [...spec.similes] : [],
	description: spec.description,
	examples: (spec.examples ?? []) as ActionExample[][],

	validate: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		options?: HandlerOptions,
	): Promise<boolean> => {
		const __avTextRaw =
			typeof message?.content?.text === "string" ? message.content.text : "";
		const __avText = __avTextRaw.toLowerCase();
		const __avKeywords = ["record", "experience"];
		const __avKeywordOk =
			__avKeywords.length > 0 &&
			__avKeywords.some((kw) => kw.length > 0 && __avText.includes(kw));
		const __avRegex = /\b(?:record|experience)\b/i;
		const __avRegexOk = Boolean(__avText.match(__avRegex));
		const __avSource = String(message?.content?.source ?? "");
		const __avExpectedSource = "";
		const __avSourceOk = __avExpectedSource
			? __avSource === __avExpectedSource
			: Boolean(__avSource || state || runtime?.agentId || runtime?.getService);
		const __avOptions = options && typeof options === "object" ? options : {};
		const __avInputOk =
			__avText.trim().length > 0 ||
			Object.keys(__avOptions as Record<string, unknown>).length > 0 ||
			Boolean(message?.content && typeof message.content === "object");

		if (!(__avKeywordOk && __avRegexOk && __avSourceOk && __avInputOk)) {
			return false;
		}

		const __avLegacyValidate = async (
			_runtime: IAgentRuntime,
			message: Memory,
		) => {
			const text = message.content.text?.toLowerCase();
			return text?.includes("remember") || text?.includes("record") || false;
		};
		try {
			return Boolean(await __avLegacyValidate(runtime, message));
		} catch {
			return false;
		}
	},

	async handler(
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		_options?: HandlerOptions,
		_callback?: HandlerCallback,
	): Promise<ActionResult> {
		void _options;
		void _callback;

		logger.info("Recording experience for message:", message.id);

		// Create experience memory with context
		const experienceMemory: Memory = {
			id: createUniqueUuid(runtime, `experience-${message.id}`),
			entityId: message.entityId,
			agentId: runtime.agentId,
			roomId: message.roomId,
			content: {
				text: message.content.text,
				source: message.content.source,
				type: "experience",
				context: state?.text,
			},
			createdAt: Date.now(),
		};

		// Store in experiences table
		await runtime.createMemory(experienceMemory, "experiences", true);
		logger.info("Experience recorded successfully");

		return {
			success: true,
			text: "Experience recorded.",
			data: {
				experienceMemoryId: experienceMemory.id,
			},
		};
	},
};
